### PR TITLE
Refactor protocol

### PR DESCRIPTION
**Pull request by KimiWu123**

### CODE REVIEW

File: zkvote/service/manager/manager.go [hunk 1]
```diff
@@ -383,73 +399,115 @@ func (m *Manager) FindProposers() (<-chan peer.AddrInfo, error) {
...
 
-// Collect ...
-func (m *Manager) Collect() (<-chan *subject.Subject, error) {
-	out := make(chan *subject.Subject, 100)
-	defer close(out)
-
-	proposers, err := m.FindProposers()
-	if err != nil {
-		fmt.Println(err)
+func (m *Manager) waitCollect(ch chan []string) {
```
Comment (juincc0) [inline zkvote/service/manager/manager.go]: `collectWorker` ? or `collectHandler` ?

File: zkvote/service/manager/manager.go [hunk 2]
```diff
@@ -355,17 +359,29 @@ func (m *Manager) Join(subjectHashHex string, identityCommitmentHex string) erro
...
 
 		// Sync ballots
-		_ = m.SyncBallotIndex(subjHex)
+		go func(ch chan bool) {
+			select {
+			case <-ch:
+			case <-time.After(10 * time.Second):
+				utils.LogWarning("wait for sync Idnetity Index timeout")
+			}
+			finished, err := m.SyncBallotIndex(subjHex)
```
Comment (juincc0) [inline zkvote/service/manager/manager.go]: will this be blocked by `select` ?

File: zkvote/service/manager/manager.go [hunk 3]
```diff
@@ -383,73 +399,115 @@ func (m *Manager) FindProposers() (<-chan peer.AddrInfo, error) {
...
 
-	var peerNum int
-	for peer := range proposers {
-		// Ignore self ID
-		if peer.ID == m.Host.ID() {
+	close(ch)
+}
+
+// Collect ...
+func (m *Manager) Collect() {
```
Comment (juinc) [inline zkvote/service/manager/manager.go]: no return value?

Comment (juincc0) [inline zkvote/service/manager/manager.go]: Rename to `SyncSubjects` ?

**Pull request merged**